source/extensions/filters/network/dubbo_proxy/message.h

```diff
@@ -53,32 +86,59 @@ enum class RpcResponseType : uint8_t {
   ResponseNullValueWithAttachments = 5,
 };
 
-class Message {
+class Context {
 public:
-  virtual ~Message() = default;
-  virtual MessageType messageType() const PURE;
-  virtual int32_t bodySize() const PURE;
-  virtual bool isEvent() const PURE;
-  virtual int64_t requestId() const PURE;
-  virtual std::string toString() const PURE;
+  using AttachmentMap = std::unordered_map<std::string, std::string>;
+
+  bool hasAttachments() const { return attachments_.empty(); }
+  const AttachmentMap& attachments() const { return attachments_; }
+
+  Buffer::Instance& message_origin_data() { return message_origin_buffer_; }
+  size_t message_size() const { return header_size() + body_size(); }
+
+  virtual size_t body_size() const PURE;
+  virtual size_t header_size() const PURE;
+
+protected:
+  Context() = default;
+  virtual ~Context() { attachments_.clear(); }
+
+  AttachmentMap attachments_;
+  Buffer::OwnedImpl message_origin_buffer_;
 };
 
-class RequestMessage : public virtual Message {
+using ContextSharedPtr = std::shared_ptr<Context>;
+
+/**
+ * RpcInvocation represent an rpc call
+ * See
+ * https://github.com/apache/incubator-dubbo/blob/master/dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcInvocation.java
+ */
+class RpcInvocation {
 public:
-  ~RequestMessage() override = default;
-  virtual SerializationType serializationType() const PURE;
-  virtual bool isTwoWay() const PURE;
+  virtual ~RpcInvocation() = default;
+
+  virtual const std::string& service_name() const PURE;
+  virtual const std::string& method_name() const PURE;
+  virtual const absl::optional<std::string>& service_version() const PURE;
+  virtual const absl::optional<std::string>& service_group() const PURE;
 };
 
-using RequestMessagePtr = std::unique_ptr<RequestMessage>;
+using RpcInvocationSharedPtr = std::shared_ptr<RpcInvocation>;
 
-class ResponseMessage : public virtual Message {
+/**
+ * RpcResult represent the result of an rpc call
+ * See
+ * https://github.com/apache/incubator-dubbo/blob/master/dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcResult.java
+ */
+class RpcResult {
 public:
-  ~ResponseMessage() override = default;
-  virtual ResponseStatus responseStatus() const PURE;
+  virtual ~RpcResult() = default;
+  virtual bool hasException() const PURE;
+  virtual bool hasValue() const PURE;
```

Comment:
where is this used?

Reply:
sorry，hasValue function is not used, I have deleted it.